### DISCUSSION TITLE
Make Slug field available by default

### DISCUSSION THREAD

**Discussion by ortimis** [payloadcms/payload#8859] [Feature Requests & Ideas]
Payload does not come with a slug field by default. You either have to build it yourself, or know where to find decent implementations that you can steal from (there are good implementations available e.g. in the website template). Depending on the desired UX, this can range from trivial to relatively complex.

Given that slugs are a highly common requirement for any CMS that is used to publish content on the web (I am aware that payload is designed to cater to needs beyond that), DX could be streamlined by simply making it available by default. Coming from other well-known CMSs, my expectation was that this would readily be available and was surprised that it was not.

Common requirements would suggest that a slug field should by default be
- unique
- indexed
- based on an entry's title
- yet editable on the fly by the user

More complex features could include
- localization
- integration with the redirects plugin (editing the slug of an already published entry would suggest adding a redirect to avoid dead links)

This suggestion might be at odds with payload's philosophy of not cluttering things too much and leaving the specifics of slug implementation to the developer. But if the goal is to grab significant market share from other established CMSs that have created the expectation of no-hustle-slugs, this might be the way to go.

**Comment by ortimis** [comment]
Happy to submit a PR if the maintainers gave me a hint on whether/how they would want this.

This could be made available as its own field type, which under the hood just uses the text field type. Alternatively, it could be a custom field component which you add and configurate just like any other custom component.

**Comment by ortimis** [comment]
For anyone dropping in, this was discussed on discord. By @paulpopus "im thinking that a storybook + a pnpx payload/ui add <component> approach like shadcn could actually be better than providing them out of a package, it would make it easier to maintain"

I think whatever works - important is that basic CMS requirements can be met with ease and no hustle

**Comment by paulpopus** [comment]
A slug field has now been released via https://github.com/payloadcms/payload/pull/14007

**Comment by jefferyto** [comment]
I'm surprised that the [`slugify`](https://github.com/payloadcms/payload/blob/v3.59.1/packages/payload/src/utilities/slugify.ts) function is the same one from the website template. It has many issues:

* Leading and trailing hyphens are not removed
* Multiple hyphens are not combined into one hyphen
* Characters with diacritical marks are removed completely instead of removing the diacritical mark only (e.g. convert `é` to `e`)

But most importantly: the assumption that only alphanumeric characters are allowed. If I run a Chinese or Arabic website, then I would expect my slugs to be in the same language.

I suggest using the `cleanForSlug` function from [@wordpress/url](https://www.npmjs.com/package/@wordpress/url) (or copying most of it), and also letting users of the slug field to override the `slugify` function.

**Reply by jacobsfletch** [reply]
Got started on that work here: https://github.com/payloadcms/payload/pull/14117.

**Comment by ortimis** [comment]
Cool!

The slug field should also be documented under https://payloadcms.com/docs/fields/overview

Other fields found in [baseFields](https://github.com/payloadcms/payload/tree/v3.59.1/packages/payload/src/fields/baseFields) are not documented either. Is there a reason for this? Perhaps it would make sense to speak of those as 'Wrapper Fields', alongside Data Fields, Presentational Fields, and Virtual Fields

**Reply by paulpopus** [reply]
Other things found in that directory are mostly internal/utilities and will not be documented for direct use for now.
